tests/test_experiment.py

```diff
@@ -1247,24 +1255,59 @@ def test_experiment_27_run_once_with_optimizer(self):
         mock_agg.create_aggregator_args.return_value = ({}, {})
         self.test_exp.set_aggregator(mock_agg)
         # Populate the Experiment with a mock Optimizer.
-        optim = create_autospec(Optimizer, instance=True)
-        self.test_exp.set_agg_optimizer(optim)
+        mock_optim = create_autospec(Optimizer, instance=True)
+        self.test_exp.set_agg_optimizer(mock_optim)
         # Disable breakpointing.
         self.test_exp.set_save_breakpoints(False)
-        # Call `Experiment.run_once`, hijacking declearn Vectors.
-        with patch("declearn.model.api.Vector.build") as patch_vector_build:
-            mock_vector = create_autospec(Vector, instance=True)
-            patch_vector_build.return_value = mock_vector
-            self.test_exp.run_once()
+        # Call `Experiment.run_once`, hijacking declearn Vectors
+        # due to the use of empty dicts of (mock) weights.
+        with patch.object(Vector, "build", new=create_autospec(Vector)):
+            result = self.test_exp.run_once()
         # Verify that expected operations occured.
-        self.assertEqual(patch_vector_build.call_count, 2)
-        optim.init_round.assert_called_once()
-        optim.step.assert_called_once()
+        self.assertEqual(result, 1, "run_once did not successfully run the round")
+        self.assertListEqual(
+            [name for name, *_ in mock_optim.method_calls],
+            ["get_aux", "set_aux", "init_round", "step"],
+            "aggregates Optimizer did not receive expected ordered calls"
+        )
+        
+        mock_job.get_received_optimizer_aux_var_from_round.assert_called_once()
+
+    def test_experiment_28_run_once_fails_from_unexpected_nodes_aux_var(self):
+        """Test that receiving auxiliary variables without an Optimizer fails."""
+        # Populate the Experiment with a mock Job outputting optimizer aux var.
+        mock_job = create_autospec(Job, instance=True)
+        mock_job.update_parameters.return_value = ("mock-path", None)
+        mock_job.get_received_optimizer_aux_var_from_round.return_value = (
+            {"module": {"node_id": {"key": "val"}}}  # mock aux-var dict
+        )
+        setattr(self.test_exp, "_job", mock_job)  # no proper Job setter
+        # Populate the Experiment with a mock Aggregator.
+        mock_agg = create_autospec(Aggregator, instance=True)
+        mock_agg.aggregator_name = "mock"
+        mock_agg.create_aggregator_args.return_value = ({}, {})
+        self.test_exp.set_aggregator(mock_agg)
+        # Disable breakpointing.
+        self.test_exp.set_save_breakpoints(False)
+        # Call run once, which should fail due to the lack of Optimizer
+        # to process the auxiliary variables received from nodes.
+        with patch.object(FedbiomedExperimentError, "__init__") as patch_exc:
+            patch_exc.return_value = None  # __init__ must return None
+            self.assertRaises(SystemExit, self.test_exp.run_once)
```

Comment:
wow this is very nice! I think we can even change other pieces of the test where we are testing that some method triggers `SystemExit`, but never the underlying exception !